lib/common/socketpool.c

```diff
@@ -229,6 +239,42 @@ void h2o_socketpool_init_target_by_hostport(h2o_socketpool_target_t *target, h2o
     } else {
         target->url = NULL;
     }
+    h2o_linklist_init_anchor(&target->_shared.sockets);
+}
+
+static h2o_socketpool_target_t *lookup_target(h2o_socketpool_t *pool, h2o_iovec_t host, uint16_t port, int is_ssl)
+{
+    /* FIXME stop using linear search and use hashtable */
+    int i = 0;
+    for (; i != pool->targets.size; ++i) {
+        h2o_socketpool_target_t *target = &pool->targets.entries[i];
+        if (target->is_ssl != is_ssl)
+            continue;
+        if (target->peer.port == 0 || target->peer.port != port)
+            continue;
+        if (memcmp(target->peer.host.base, host.base, host.len) != 0)
+            continue;
+        return target;
+    }
+    return NULL;
+}
+
+h2o_socketpool_target_t *h2o_socketpool_get_or_add_target(h2o_socketpool_t *pool, h2o_iovec_t host, uint16_t port, int is_ssl, h2o_url_t *url)
```

Comment:
Current design seems to expose a function that searches for or adds a target (i.e. `get_or_add_target`) and passes the address of the target to the function that actually connects to the server.

The approach seem to have two issues:
* once allocated, `pool->targets` becomes unexpandable (or shrinkable), unless the mutex is locked from before the call to `get_or_add_target` to after the call to `connect_to_target`. Otherwise, we would have a race condition.
* a target entry would be created for a server that always sends `connection: close`, which is IMO a waste

So you might want to consider the following:
* do not expose `get_or_add_target`
* accept the (host, port, is_ssl) triplet as an argument to `connect_to_target`
 * the function checks if a matching target exists, and if a target with a pooled socket found, returns the socket
 * otherwise, the socket pool connects to a new socket (note that you do not need to ensure that a matching target exists)
* in `h2o_socketpool_return`, look for a matching target, and if _not_ found, create a new target. Add to the target the socket being returned. You might need to pass the triplet as arguments to the function.

By using this approach, all locking operations can be kept within the socketpool.c. Also, `pool->targets` becomes not only expandable but also shrinkable (i.e. targets without any poold sockets can be garbage-collected).